### PR TITLE
Updating the purchase details from the database

### PR DESCRIPTION
**Pull request by EXR-85**

### CODE REVIEW

File: src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java [hunk 1]
```diff
@@ -38,11 +38,11 @@ public ResponseEntity<?> deletePurchase(@PathVariable("purchaseId") UUID purchas
         try {
             boolean success = purchaseService.deletePurchaseDetails(purchaseId);
             deleteResponse.setSuccess(success);
-            deleteResponse.setErrorMessages(ApplicationConstants.DELETED);
+            deleteResponse.setErrorMessage(ApplicationConstants.SUCCESSFULLY_DELETED);
             httpStatus = HttpStatus.OK;
         } catch (NoRecordFoundException e) {
```
Comment (mh-sraveendran) [inline src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java]: Follow the same pattern as in DeletionController for handling NoRecordFoundException.
This method just throws that exception. Its handled in GLobalExeptionHandler

Comment (EXR-85) [reply src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java]: Updated.

File: src/main/java/com/elixr/poc/rest/controller/PurchaseModificationController.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * Calling updatePurchase method with the parameter purchaseId to update the purchase by purchaseId.
+     * And handling the Exception if the userId is not matching.
+     *
+     * @param purchaseId
+     * @return
+     * @throws NoRecordFoundException
+     */
+    @PatchMapping("/purchase/{purchaseId}")
+    public ResponseEntity<Purchase> updateUserPartially(@PathVariable("purchaseId") UUID purchaseId, @RequestBody @Valid Purchase purchaseDetails) throws NoRecordFoundException {
```
Comment (mh-sraveendran) [inline src/main/java/com/elixr/poc/rest/controller/PurchaseModificationController.java]: Dont use Purchase class in argument or return type of a Controller.
use PurchaseReuqest & PurchaseResponse. 
do the conversion from request/response object to Purchase object in controller

File: src/main/java/com/elixr/poc/service/PurchaseService.java [hunk 3]
```diff
@@ -33,4 +34,21 @@ public boolean deletePurchaseDetails(UUID purchaseId) throws NoRecordFoundExcept
...
+    /**
+     * If the PurchaseId exists then update the PurchaseId or else send Id is mismatched.
+     *
+     * @param purchaseId
+     * @return
+     * @throws NoRecordFoundException
+     */
+    public Purchase updateUserPartially(UUID purchaseId, Purchase purchaseDetails) throws NoRecordFoundException {
+        Purchase purchase = purchaseRepository.findById(purchaseId).orElseThrow(() -> new NoRecordFoundException("User not found on :: " + purchaseId));
+        purchase.setUserName(purchaseDetails.getUserName());
```
Comment (mh-sraveendran) [inline src/main/java/com/elixr/poc/service/PurchaseService.java:67]: add validation.
throw an error if username  does not exists in user table

File: src/main/java/com/elixr/poc/rest/controller/PurchaseModificationController.java (hunk 2 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/rest/controller/PurchaseModificationController.java]: Updated.

File: src/main/java/com/elixr/poc/constants/ApplicationConstants.java [hunk 4]
```diff
@@ -3,20 +3,20 @@
...
  * This constant file holds all the error messages for all the end point
  */
+
 import lombok.AccessLevel;
 import lombok.NoArgsConstructor;
 
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
 public class ApplicationConstants {
 
+    public static final String AMOUNT_MISSING_IN_PURCHASE = "Amount is mandatory";
```
Comment (mh-spillai) [inline src/main/java/com/elixr/poc/constants/ApplicationConstants.java]: Its better to add validation messages in the message.properties file.

File: src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java [hunk 5]
```diff
@@ -32,19 +32,12 @@ public PurchaseDeletionController(PurchaseService purchaseService) {
...
-            deleteResponse.setSuccess(success);
-            deleteResponse.setErrorMessages(ApplicationConstants.DELETED);
-            httpStatus = HttpStatus.OK;
-        } catch (NoRecordFoundException e) {
-            deleteResponse.setSuccess(false);
-            deleteResponse.setErrorMessages(e.getMessage());
-            httpStatus = HttpStatus.BAD_REQUEST;
-        }
+        boolean success = purchaseService.deletePurchaseDetails(purchaseId);
+        deleteResponse = CommonErrorResponse.builder().success(success).errorMessage(ApplicationConstants.SUCCESSFULLY_DELETED).build();
```
Comment (mh-spunnappala) [inline src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java]: Why is the response object name as ErrorResponse, even if it is a success? Better to rename the class and attributes of the class, or use another success response class if you have

File: src/main/java/com/elixr/poc/service/PurchaseService.java [hunk 6]
```diff
@@ -20,17 +22,34 @@ public PurchaseService(PurchaseRepository purchaseRepository) {
...
-        boolean success = false;
+    public boolean deletePurchaseDetails(UUID purchaseId) throws GlobalException {
+        boolean success;
         boolean purchaseRecordExists = purchaseRepository.existsById(purchaseId);
         if (purchaseRecordExists) {
             purchaseRepository.deleteById(purchaseId);
             success = true;
         } else {
-            throw new NoRecordFoundException(ApplicationConstants.ID_MISMATCH);
+            throw new GlobalException(ApplicationConstants.ID_MISMATCH);
```
Comment (mh-spunnappala) [inline src/main/java/com/elixr/poc/service/PurchaseService.java]: Error message not matching with the specific use case. Here the scenario is the id does not exist, not the id mismatch.

**Review by mh-spunnappala — CHANGES_REQUESTED** [review]

File: src/main/java/com/elixr/poc/constants/ApplicationConstants.java (hunk 4 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/constants/ApplicationConstants.java]: Updated

File: src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java (hunk 5 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java]: Updated has per your suggestion

File: src/main/java/com/elixr/poc/service/PurchaseService.java (hunk 6 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/service/PurchaseService.java]: Updated

File: src/main/java/com/elixr/poc/common/util/MessagesUtil.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    private static final String MESSAGE_PROPERTY_FILE_PREFIX = "messages";
+
+    private static ResourceBundle moduleMessagesResourceBundle;
+
+    static {
+        try {
+            // Initialize messagesResourceBundle in a static block because the module may not have messages.properties.
+            moduleMessagesResourceBundle = ResourceBundle.getBundle(MESSAGE_PROPERTY_FILE_PREFIX);
+        } catch (MissingResourceException e) {
+            moduleMessagesResourceBundle = null;
```
Comment (kssreejith1994) [inline src/main/java/com/elixr/poc/common/util/MessagesUtil.java]: A catch block without an exception rethrowed or no logs should never be used. You will have no way of knowing what is the error in your program.

File: src/main/java/com/elixr/poc/common/util/MessagesUtil.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        }
+    }
+
+    public static String getMessage(String key, Object... arguments) {
```
Comment (kssreejith1994) [inline src/main/java/com/elixr/poc/common/util/MessagesUtil.java]: Ca you explain what does Object... means?

File: src/main/java/com/elixr/poc/exception/GlobalExceptionHandler.java [hunk 9]
```diff
@@ -17,17 +19,44 @@ public class GlobalExceptionHandler {
...
+     * Generic exceptions are handled.
+     *
+     * @param exception
+     * @return
+     */
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity handleGenericException(Exception exception) {
+        CommonResponse commonResponse = CommonResponse.builder().success(false)
+                .errorMessage(exception.getLocalizedMessage()).build();
+        return new ResponseEntity<>(commonResponse, HttpStatus.BAD_REQUEST);
```
Comment (kssreejith1994) [inline src/main/java/com/elixr/poc/exception/GlobalExceptionHandler.java]: Can you try to create an internal server error scenario and check what is the http response code getting returned? Please post the response you get here.

File: src/main/java/com/elixr/poc/rest/request/PurchaseRequest.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package com.elixr.poc.rest.request;
+
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+
+@Data
+public class PurchaseRequest {
```
Comment (kssreejith1994) [inline src/main/java/com/elixr/poc/rest/request/PurchaseRequest.java:7]: How are the error messages handled?

Comment (kssreejith1994) [reply src/main/java/com/elixr/poc/rest/request/PurchaseRequest.java:7]: If the fields are not passed.

File: src/main/java/com/elixr/poc/rest/response/CommonResponse.java [hunk 11]
```diff
@@ -1,9 +1,11 @@
 package com.elixr.poc.rest.response;
 
+import lombok.Builder;
 import lombok.Data;
 
 @Data
-public class DeleteResponse {
+@Builder
+public class CommonResponse {
```
Comment (kssreejith1994) [inline src/main/java/com/elixr/poc/rest/response/CommonResponse.java]: What is this response used for? You already have another common response class?

File: src/main/java/com/elixr/poc/service/PurchaseService.java [hunk 12]
```diff
@@ -20,17 +23,35 @@ public PurchaseService(PurchaseRepository purchaseRepository) {
      *
      * @param purchaseId
      * @return
-     * @throws NoRecordFoundException
+     * @throws IdNotFoundException
      */
-    public boolean deletePurchaseDetails(UUID purchaseId) throws NoRecordFoundException {
-        boolean success = false;
+    public boolean deletePurchaseDetails(UUID purchaseId) throws IdNotFoundException {
```
Comment (kssreejith1994) [inline src/main/java/com/elixr/poc/service/PurchaseService.java]: You are already handling the IdNotFoundException. Why is it throwed again by the method?

File: src/main/java/com/elixr/poc/service/PurchaseService.java [hunk 13]
```diff
@@ -20,17 +23,35 @@ public PurchaseService(PurchaseRepository purchaseRepository) {
...
+
+    /**
+     * If the PurchaseId exists then update the PurchaseId or else send Id is mismatched.
+     *
+     * @param purchaseId
+     * @return
+     * @throws IdNotFoundException
+     */
+    public Purchase purchaseUpdate(UUID purchaseId, PurchaseRequest purchaseDetails ) throws IdNotFoundException {
+        Purchase purchase = purchaseRepository.findById(purchaseId).orElseThrow(() -> new IdNotFoundException("Purchase not found on :: " + purchaseId));
```
Comment (kssreejith1994) [inline src/main/java/com/elixr/poc/service/PurchaseService.java]: Please explain how does this line work?

**Review by kssreejith1994 — CHANGES_REQUESTED** [review]

File: src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java [hunk 14]
```diff
@@ -1,50 +1,45 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.UUID;
+
+@RestController
+@RequestMapping("/controller")
```
Comment (mh-ppremlatha) [inline src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java]: Why do you have /controller in the path? It does not match the requirement

File: src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java [hunk 15]
```diff
@@ -1,50 +1,45 @@
...
+     * @param purchaseId
+     * @return
+     */
+    @DeleteMapping("/purchase/{purchaseId}")
+    public ResponseEntity deletePurchase(@PathVariable("purchaseId") UUID purchaseId) throws IdNotFoundException {
+        CommonResponse deleteResponse;
+        HttpStatus httpStatus;
+        boolean success = purchaseService.deletePurchaseDetails(purchaseId);
+        deleteResponse = CommonResponse.builder().success(success)
+                .errorMessage(MessagesUtil.getMessage(MessagesKeyEnum.ENTITY_DELETED_SUCCESSFULLY.getKey())).build();
```
Comment (mh-ppremlatha) [inline src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java]: Why are you passing success message in the 'errorMessage' field?

File: src/main/java/com/elixr/poc/rest/controller/PurchaseModificationController.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.UUID;
+
+@RestController
+@Validated
+@RequestMapping("/controller")
```
Comment (mh-ppremlatha) [inline src/main/java/com/elixr/poc/rest/controller/PurchaseModificationController.java]: Why do you have /controller in the path? It does not match the requirement

File: src/main/java/com/elixr/poc/rest/response/CommonResponse.java [hunk 17]
```diff
@@ -1,9 +1,11 @@
...
     private boolean success;
     private String errorMessage;
```
Comment (mh-ppremlatha) [inline src/main/java/com/elixr/poc/rest/response/CommonResponse.java]: Why do you need this field here? You have a CommonErrorMessage class with errorMessage field?

**Review by mh-ppremlatha — CHANGES_REQUESTED** [review]

File: src/main/java/com/elixr/poc/common/util/MessagesUtil.java (hunk 7 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/common/util/MessagesUtil.java]: Updated.

File: src/main/java/com/elixr/poc/rest/response/CommonResponse.java (hunk 11 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/rest/response/CommonResponse.java]: Made changes.

File: src/main/java/com/elixr/poc/service/PurchaseService.java (hunk 12 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/service/PurchaseService.java]: Removed

File: src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java (hunk 14 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java]: Removed.

File: src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java (hunk 15 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/rest/controller/PurchaseDeletionController.java]: Updated

File: src/main/java/com/elixr/poc/rest/controller/PurchaseModificationController.java (hunk 16 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/rest/controller/PurchaseModificationController.java]: Removed.

File: src/main/java/com/elixr/poc/rest/response/CommonResponse.java (hunk 17 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/rest/response/CommonResponse.java]: Updated.

File: src/main/java/com/elixr/poc/common/util/MessagesUtil.java (hunk 8 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/common/util/MessagesUtil.java]: Object... is used for multiple arguments to be passed.

File: src/main/java/com/elixr/poc/service/PurchaseService.java (hunk 3 above)
Comment (EXR-85) [reply src/main/java/com/elixr/poc/service/PurchaseService.java:67]: Updated.

**Review by mh-sraveendran — APPROVED** [review]

**Pull request merged**